sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/models/BlobDownloadAsyncResponse.java

```diff
@@ -0,0 +1,112 @@
+        Function<HttpGetterInfo, Mono<BlobDownloadAsyncResponse>> getter) {
+        super(request, statusCode, headers, value, deserializedHeaders);
+
+        StorageImplUtils.assertNotNull("getter", getter);
+        StorageImplUtils.assertNotNull("info", info);
+        StorageImplUtils.assertNotNull("info.eTag", info.getETag());
+        this.info = info;
+        this.getter = getter;
+        this.options = (options == null) ? new ReliableDownloadOptions() : options;
+    }
+
+    @Override
+    public Flux<ByteBuffer> getValue() {
+        /*
+        We pass -1 for currentRetryCount because we want tryContinueFlux to receive a value of 0 for number of
+        retries as we have not actually retried yet, only made the initial try. Because applyReliableDownload() will
+        add 1 before calling into tryContinueFlux, we set the initial value to -1.
+         */
+        Flux<ByteBuffer> value = (options.maxRetryRequests() == 0)
+            ? super.getValue()
+            : applyReliableDownload(super.getValue(), -1, options);
+
+        return value.switchIfEmpty(Flux.just(ByteBuffer.wrap(new byte[0])));
+    }
+
+    private Flux<ByteBuffer> tryContinueFlux(Throwable t, int retryCount, ReliableDownloadOptions options) {
+        // If all the errors are exhausted, return this error to the user.
+        if (retryCount > options.maxRetryRequests() || !(t instanceof IOException)) {
+            return Flux.error(t);
+        } else {
+            /*
+            We wrap this in a try catch because we don't know the behavior of the getter. Most errors would probably
+            come from an unsuccessful request, which would be propagated through the onError methods. However, it is
+            possible the method call that returns a Single is what throws (like how our apis throw some exceptions at
+            call time rather than at subscription time.
+             */
+            try {
+                /*Get a new response and try reading from it.
+                Do not compound the number of retries by passing in another set of downloadOptions; just get
+                the raw body.
+                */
+                return getter.apply(info)
+                    .flatMapMany(response -> applyReliableDownload(super.getValue(), retryCount, options));
+            } catch (Exception e) {
+                // If the getter fails, return the getter failure to the user.
+                return Flux.error(e);
+            }
+        }
+    }
+
+    private Flux<ByteBuffer> applyReliableDownload(Flux<ByteBuffer> data, int currentRetryCount,
+        ReliableDownloadOptions options) {
+        return data.doOnNext(buffer -> {
+            /*
+            Update how much data we have received in case we need to retry and propagate to the user the data we
+            have received.
+             */
+            this.info.setOffset(this.info.getOffset() + buffer.remaining());
```

Comment:
It's possible to encounter a race condition here because we are not atomically doing this get and update set. And flatMapMany can be run asynchronously.

Reply:
It's my understanding that the contract of reactive streams states that calls to onNext must be strictly sequential; no two threads can enter the method at the same time. If that is the case, there is no danger of race conditions. Running asynchronously does not necessarily imply that anything is happening in parallel.

Reply:
I thought having side effects would just wait for the subscriber to do its onNext computation but not necessarily the side effect.

Looks like this is isn't a problem.
https://stackoverflow.com/a/53748683

Reply:
My understanding of .doOnNext were for operations that had no consequence like logging. But we are processing the element here, something that map or flatmap is better suited for.
https://stackoverflow.com/a/44372692